keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,802 @@
+have to override these values in case they are affected by supplying the flags
+with updated defaults. A configuration format like the one proposed can be
+versioned over time with changing defaults, mitigating this risk by only
+affecting users who knowingly bump the version.
+
+#### Story 6: Conditionally filtering requests to webhooks
+
+Let's say if an API request is being made by an user in a specific group, the
+webhook request can be skipped.
+
+
+### Risks and Mitigations
+
+- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
+each are configured. This may create inconsistencies. Mitigation is to have the
+cluster bootstrapper handle such scenarios.
+- In case an administrator enables this feature and the webhook kubeconfig file
+is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
+will not be able to start. This can be mitigated by fixing the malformed values.
+
+
+## Design Details
+
+We would like to introduce a structured file format which allows authorization
+to be configured using a flag (`--authorization-config-file`) which accepts a
+path to a file on the disk. Setting both `--authorization-config-file` and
+configuring an authorization webhook using the `--authorization-webhook-*`
+command line flags will not be allowed. If the user does that,
+there will be an error and API Server would exit right away.
+
+The configuration would be validated at startup and the API server will fail to
+start if the configuration is invalid.
+
+The API server will periodically reload the configuration at a specific time
+interval. If it changes, the new configuration will be used for the Authorizer
+chain. The reloader will also check if the webhook is healthy, thereby
+preventing any typo/misconfiguration with the Webhook resulting in bad
+Authorizer config. If healthcheck on the webhook failed, the last known good
+config will be used. In the next iteration of reload, if webhook is found to be
+healthy, the new config will be used. Logging and metrics would be used to
+signal success/failure of a config reload so that cluster admins can have
+observability over this process. Reload must not add or remove Node or RBAC
+authorizers. They can be reordered, but cannot be added or removed.
+
+The proposed structure is illustrated below:
+
+> The sample configuration describes all the fields, their defaults and possible
+values.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthorizationConfiguration
+authorizers:
+  - type: Webhook
+    webhook:
+      # Name used to describe the webhook
+      # This is explicitly used in monitoring machinery for metrics
+      # Note:
+      #   - Do exercise caution when setting the value
+      #   - If not specified, the default would be set to ""
+      #   - If there are multiple webhooks in the authorizer chain,
+      #     this field is required
+      name: super-important-kube-system-authorizer
```

Comment:
This `name` field is new, for existing webhooks configured via the command line flag, what value would we use?

Reply:
@enj what are your opinions on the above?

Reply:
`my_fancy_webhook`?  The value doesn't really matter because it is meant as a human friendly name.  But I do not really see value is allowing it to be optional.